### PR TITLE
JAX-WS + JAX-RS 2.1: Upgrade to CXF 3.5.5

### PR DESCRIPTION
**Pull request by neuwerk**
This pull request does the following:

1.) Updates the following binaries (and removes the 3.4.3 binaries as dependencies): 

```
org.apache.cxf:cxf-core:3.5.5
org.apache.cxf:cxf-rt-bindings-soap:3.5.5
org.apache.cxf:cxf-rt-bindings-xml:3.5.5
org.apache.cxf:cxf-rt-databinding-jaxb:3.5.5
org.apache.cxf:cxf-rt-features-logging:3.5.5
org.apache.cxf:cxf-rt-frontend-jaxrs:3.5.5
org.apache.cxf:cxf-rt-frontend-jaxws:3.5.5
org.apache.cxf:cxf-rt-frontend-simple:3.5.5
org.apache.cxf:cxf-rt-management:3.5.5
org.apache.cxf:cxf-rt-rs-client:3.5.5
org.apache.cxf:cxf-rt-rs-extension-reactivestreams:3.5.5
org.apache.cxf:cxf-rt-rs-extension-rx2:3.5.5
org.apache.cxf:cxf-rt-rs-mp-client:3.5.5
org.apache.cxf:cxf-rt-security-saml:3.5.5
org.apache.cxf:cxf-rt-security:3.5.5
org.apache.cxf:cxf-rt-transports-http-hc:3.5.5
org.apache.cxf:cxf-rt-transports-http:3.5.5
org.apache.cxf:cxf-rt-ws-addr:3.5.5
org.apache.cxf:cxf-rt-ws-mex:3.5.5
org.apache.cxf:cxf-rt-ws-policy:3.5.5
org.apache.cxf:cxf-rt-ws-security:3.5.5
org.apache.cxf:cxf-rt-wsdl:3.5.5
org.apache.cxf:cxf-tools-common:3.5.5
org.apache.cxf:cxf-tools-validator:3.5.5
org.apache.cxf:cxf-tools-wadlto-jaxrs:3.5.5
org.apache.santuario:xmlsec:2.2.3
org.apache.wss4j:wss4j-bindings:2.4.1
org.apache.wss4j:wss4j-policy:2.4.1
org.apache.wss4j:wss4j-ws-security-common:2.4.1
org.apache.wss4j:wss4j-ws-security-dom:2.4.1
org.apache.wss4j:wss4j-ws-security-policy-stax:2.4.1
org.apache.wss4j:wss4j-ws-security-stax:2.4.1
org.apache.wss4j:wss4j-ws-security-web:2.4.1
org.ehcache:ehcache:3.9.3
```

2.)  Updates code overriding in Open Liberty to clearly indicate changes made to the original source by the Liberty teams
3.) Adds clarifying details about overrides where possible. 
3.) Updates code overrides to include changes made to source files between the versions of CXF.

### CODE REVIEW

**Comment by neuwerk** [comment]
#build

**Comment by neuwerk** [comment]
#build

**Comment by neuwerk** [comment]
#build

**Comment by neuwerk** [comment]
#build

**Comment by neuwerk** [comment]
#build
#spawn.full.fats=com.ibm.ws.jaxws.2.2.webcontainer_fat,com.ibm.ws.jaxws.2.2.webcontainer_fat_extended, com.ibm.ws.jaxws.clientcontainer_fat,com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.ejb_fat,com.ibm.ws.jaxws.X.wsat_fat, com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor, io.openliberty.ws.jaxrs.2.1.jaxws.2.2_fat,io.openliberty.ws.jaxrs.global.handler.internal_fat, io.openliberty.ws.jaxws.global.handler.internal_fat,com.ibm.ws.wsat.common_fat,com.ibm.ws.wsat.concurrent_fat, com.ibm.ws.wsat.migration_fat.1,com.ibm.ws.wsat.migration_fat.2,com.ibm.ws.wsat.migration_fat.3,com.ibm.ws.wsat.migration_fat.4,com.ibm.ws.wsat.migration_fat.5,com.ibm.ws.wsat.recovery_fat.multi.1,com.ibm.ws.wsat.recovery_fat.multi.2,com.ibm.ws.wsat.recovery_fat.multi.3,com.ibm.ws.wsat.recovery_fat.multi.4,com.ibm.ws.wsat.recovery_fat.single,com.ibm.ws.wsat_fat.db,com.ibm.ws.wsat_fat.ssl,com.ibm.ws.wsat_fat.multi,com.ibm.ws.wsat_fat.assertion,com.ibm.ws.wsat_fat.dbs_optional,com.ibm.ws.wsat_fat.dbs,com.ibm.ws.wsat_fat.db_optional,com.ibm.ws.wssecurity_fat.wsscxf.saml.3,com.ibm.ws.wssecurity_fat.wsscxf.1,com.ibm.ws.wssecurity_fat.wsscxf.saml.2,com.ibm.ws.wssecurity_fat.wsscxf.saml.4,com.ibm.ws.wssecurity_fat.wsscxf,com.ibm.ws.wssecurity_fat.wsscxf.saml.1,com.ibm.ws.wssecurity_fat.wsscxf.saml,com.ibm.ws.wssecurity_fat.wsscxf.saml.5,com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps

File: dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/bnd.bnd [hunk 1]
```diff
@@ -26,6 +26,7 @@
 # (yes, unused rules cause failures...). So, without a co-delivery
 # (changing the OL code at the same time we change the WS-CD-Open code),
 # it is not possible to update the export version.
+importVer=3.5.5
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/bnd.bnd:29]: Please consider updating the version numbers in the comments above this.

File: dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/bnd.bnd [hunk 2]
```diff
@@ -157,7 +158,7 @@ DynamicImport-Package: \
 
 # CXF's version is 3.4.3 even tho exportVer=3.2.4 see top comment
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/bnd.bnd:159]: Remove this comment

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/bnd.bnd [hunk 3]
```diff
@@ -4,7 +4,7 @@
 # are made available under the terms of the Eclipse Public License 2.0
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/bnd.bnd:4]: Update Copyright to 2023.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/bnd.bnd [hunk 4]
```diff
@@ -15,7 +15,8 @@
 -sub: *.bnd
 
 bVersion=1.0
-cxfVersion=3.4.3
+cxfImportVersion=3.5.5
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/bnd.bnd:18]: Might be useful to add a comment to point to the comments in core.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.xml.3.2/bnd.bnd [hunk 5]
```diff
@@ -4,7 +4,7 @@
 # are made available under the terms of the Eclipse Public License 2.0
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.xml.3.2/bnd.bnd:4]: add 2023

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/JAXRSInvoker.java [hunk 6]
```diff
@@ -251,7 +251,7 @@ public Object invoke(Exchange exchange, Object request, Object resourceObject) {
 
                 result = checkSubResultObject(result, subResourcePath);
 
-                Class<?> subResponseType = null;
+                final Class<?> subResponseType;
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/JAXRSInvoker.java:254]: Is it safe to assume that this is a change in CXF and not a Liberty change?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/JAXRSInvoker.java [hunk 7]
```diff
@@ -380,7 +380,7 @@ protected AsyncResponseImpl checkFutureResponse(Message inMessage, Object result
     protected Method getMethodToInvoke(ClassResourceInfo cri, OperationResourceInfo ori, Object resourceObject) {
         Method resourceMethod = cri.getMethodDispatcher().getMethod(ori);
 
-        Method methodToInvoke = null;
+        Method methodToInvoke;
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/JAXRSInvoker.java:383]: CXF not Liberty change?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/impl/ContainerRequestContextImpl.java [hunk 8]
```diff
@@ -123,7 +133,10 @@ public void setRequestUri(URI requestUri) throws IllegalStateException {
 
     protected void doSetRequestUri(URI requestUri) throws IllegalStateException {
         checkNotPreMatch();
-        //Liberty change - using requestUri.toString() to pass CTS
+		// TODO: The JAX-RS TCK requires the full uri toString() rather than just the raw path, but
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/impl/ContainerRequestContextImpl.java:136]: I guess I'd prefer that an issue was created to investigate this rather than a TODO in the code.  Just a preference. And are we sure this change is still needed considering the additions below it?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/JAXRSOutInterceptor.java [hunk 9]
```diff
@@ -446,10 +447,10 @@ private void setResponseDate(MultivaluedMap<String, Object> headers, boolean fir
...
             return;
         }
-// Liberty Change for CXF Begin
-//        SimpleDateFormat format = HttpUtils.getHttpDateFormat();
-        headers.putSingle(HttpHeaders.DATE, CachedTime.getCachedTime().getTimeAsString(System.currentTimeMillis()));
-//Liberty Change for CXF End
+		// Liberty Change Start - Pull from chached time
+		// SimpleDateFormat format = HttpUtils.getHttpDateFormat();
+        headers.putSingle(HttpHeaders.DATE, CachedTime.getCachedTime().getTimeAsString(-1));
+		//Liberty Change End
```
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/JAXRSOutInterceptor.java:453]: Curious:  Why was this changed to -1?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.hc.3.2/src/org/apache/cxf/transport/http/asyncclient/AsyncHTTPConduit.java [hunk 10]
```diff
@@ -573,7 +570,10 @@ protected void verifySession(final HttpHost host,
                                     setSSLSession(sslsession);
                                 }
                             });
+                    // See please https://issues.apache.org/jira/browse/HTTPASYNC-168, the attribute names
+                    // are case-sensitive.
                     ctx.setAttribute("http.iosession-factory-registry", regBuilder.build());
+                    ctx.setAttribute("http.ioSession-factory-registry", regBuilder.build());
```
Comment (BerksanAtes) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.hc.3.2/src/org/apache/cxf/transport/http/asyncclient/AsyncHTTPConduit.java:576]: Is this a duplicate line?

File: dev/com.ibm.ws.org.apache.cxf.rt.ws.security.3.4.1/src/org/apache/cxf/ws/security/cache/CXFEHCacheReplayCache.java [hunk 11]
```diff
@@ -52,6 +53,7 @@ public CXFEHCacheReplayCache(String key, Bus bus, Path diskstorePath) throws WSS
     //Liberty Change
     public CXFEHCacheReplayCache(String key, Bus bus, Path diskstorePath, HashMap oldconfig) throws WSSecurityException {
         super(key, diskstorePath, oldconfig);
+		// Liberty CHangte End
```
Comment (BerksanAtes) [inline dev/com.ibm.ws.org.apache.cxf.rt.ws.security.3.4.1/src/org/apache/cxf/ws/security/cache/CXFEHCacheReplayCache.java:56]: Typo

**Review by BerksanAtes — APPROVED** [review]
Commented out 2 things that won't change functionality.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/impl/ContainerRequestContextImpl.java (hunk 8 above)
Comment (WhiteCat22) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/impl/ContainerRequestContextImpl.java:136]: agreed, we should reference the issue in the comment

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/CachedTime.java [hunk 12]
```diff
@@ -7,36 +7,28 @@
  * 
  * SPDX-License-Identifier: EPL-2.0
  *
- * Contributors:
- *     IBM Corporation - initial API and implementation
  *******************************************************************************/
-// Liberty Change for CXF Begin
+// Liberty Change for CXF Begin - Entire class is unique to our fork, could it be contributed back?
```
Comment (WhiteCat22) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/CachedTime.java:11]: We should open an issue for this

**Review by WhiteCat22 — APPROVED** [review]

**Comment by neuwerk** [comment]
#build
#spawn.full.fats=com.ibm.ws.jaxws.2.2.webcontainer_fat,com.ibm.ws.jaxws.2.2.webcontainer_fat_extended, com.ibm.ws.jaxws.clientcontainer_fat,com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.ejb_fat,com.ibm.ws.jaxws.X.wsat_fat, com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor, io.openliberty.ws.jaxrs.2.1.jaxws.2.2_fat,io.openliberty.ws.jaxrs.global.handler.internal_fat, io.openliberty.ws.jaxws.global.handler.internal_fat,com.ibm.ws.wsat.common_fat,com.ibm.ws.wsat.concurrent_fat, com.ibm.ws.wsat.migration_fat.1,com.ibm.ws.wsat.migration_fat.2,com.ibm.ws.wsat.migration_fat.3,com.ibm.ws.wsat.migration_fat.4,com.ibm.ws.wsat.migration_fat.5,com.ibm.ws.wsat.recovery_fat.multi.1,com.ibm.ws.wsat.recovery_fat.multi.2,com.ibm.ws.wsat.recovery_fat.multi.3,com.ibm.ws.wsat.recovery_fat.multi.4,com.ibm.ws.wsat.recovery_fat.single,com.ibm.ws.wsat_fat.db,com.ibm.ws.wsat_fat.ssl,com.ibm.ws.wsat_fat.multi,com.ibm.ws.wsat_fat.assertion,com.ibm.ws.wsat_fat.dbs_optional,com.ibm.ws.wsat_fat.dbs,com.ibm.ws.wsat_fat.db_optional,com.ibm.ws.wssecurity_fat.wsscxf.saml.3,com.ibm.ws.wssecurity_fat.wsscxf.1,com.ibm.ws.wssecurity_fat.wsscxf.saml.2,com.ibm.ws.wssecurity_fat.wsscxf.saml.4,com.ibm.ws.wssecurity_fat.wsscxf,com.ibm.ws.wssecurity_fat.wsscxf.saml.1,com.ibm.ws.wssecurity_fat.wsscxf.saml,com.ibm.ws.wssecurity_fat.wsscxf.saml.5,com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps

**Comment by neuwerk** [comment]
#libby

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/CachedTime.java (hunk 12 above)
Comment (jhanders34) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/CachedTime.java:11]: Yeah in the PR I delivered to update CachedTime, it was asked for me to contribute to cxf.  I hadn't gotten to it, but would be willing to do so.

Comment (jhanders34) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/CachedTime.java:11]: Why did you take my changes out.  The changes in this class should be undone in a new PR.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/JAXRSOutInterceptor.java (hunk 9 above)
Comment (jhanders34) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/JAXRSOutInterceptor.java:453]: Yeah this should be undone just like the CachedTime change should be undone.  This is completely undoing my last PR on this.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/test/com/ibm/ws/jaxrs/utils/test/JAXRSOutInterceptorTest.java [hunk 13]
```diff
@@ -18,16 +18,14 @@
 import java.util.Date;
 import java.util.Locale;
 
-import org.apache.cxf.jaxrs.interceptor.CachedTimeAccessor;
 import org.apache.cxf.jaxrs.utils.HttpUtils;
 import org.junit.Test;
 
-import junit.framework.Assert;
-
+/**
```
Comment (jhanders34) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/test/com/ibm/ws/jaxrs/utils/test/JAXRSOutInterceptorTest.java:24]: Same here.  My last PR is completely undone.  I can make the PR to put my change back if you want.
